src/code/sched.c

```diff
@@ -1,47 +1,602 @@
+            gRSPAudioTotalTime += time - sRSPAudioStartTime;
+            break;
+        case M_GFXTASK:
+            sRSPGFXTotalTime += time - sRSPGFXStartTime;
+            break;
+        default:
+            if (1) {}
+            sRSPOtherTotalTime += time - sRSPOtherStartTime;
+            break;
+    }
+
+    curRSP = sched->curRSPTask;
+    sched->curRSPTask = NULL;
+
+    if (curRSP->list.t.type == M_AUDTASK) {
+        // Reset the global audio task data pointers
+        gAudioSPDataPtr = NULL;
+        gAudioSPDataSize = 0;
+    }
+
+    if ((curRSP->state & OS_SC_YIELD) && osSpTaskYielded(&curRSP->list)) {
+        // If the task was yielded, re-queue the task
+        curRSP->state |= OS_SC_YIELDED;
+        curRSP->next = sched->gfxListHead;
+        sched->gfxListHead = curRSP;
+        if (sched->gfxListTail == NULL) {
+            sched->gfxListTail = curRSP;
+        }
+    } else {
+        // Mark task completed
+        curRSP->state &= ~OS_SC_SP;
+        Sched_NotifyDone(sched, curRSP);
+    }
+
+    // Schedule and run next task
+    state = ((sched->curRSPTask == NULL) << 1) | (sched->curRDPTask == NULL);
+    if (Sched_Schedule(sched, &nextRSP, &nextRDP, state) != state) {
+        Sched_RunTask(sched, nextRSP, nextRDP);
+    }
+}
+
+void Sched_HandleRDPDone(SchedContext* sched) {
+    OSScTask* curRDP;
+    OSScTask* nextRSP = NULL;
+    OSScTask* nextRDP = NULL;
+    s32 state;
+
+    if (sched->curRDPTask == NULL) {
+        osSyncPrintf("__scHandleRDP:sc->curRDPTask == NULL\n");
+        return;
+    }
+
+    // Log run time
+    gRDPTotalTime = osGetTime() - sRDPStartTime;
+
+    // Mark task done
+    curRDP = sched->curRDPTask;
+    sched->curRDPTask = NULL;
+    curRDP->state &= ~OS_SC_DP;
+
+    Sched_NotifyDone(sched, curRDP);
+
+    // Schedule and run next task
+    state = ((sched->curRSPTask == NULL) << 1) | (sched->curRDPTask == NULL);
+    if (Sched_Schedule(sched, &nextRSP, &nextRDP, state) != state) {
+        Sched_RunTask(sched, nextRSP, nextRDP);
+    }
+}
+
+/**
+ * Sends a message to the scheduler to inform it that a new task has
+ * been sent down the command queue.
+ */
+void Sched_SendEntryMsg(SchedContext* sched) {
+    osSendMesg(&sched->interruptQ, ENTRY_MSG, OS_MESG_BLOCK);
+}
+
+/**
+ * Sends a message to the scheduler to inform it that it should attempt
+ * to stop the last dispatched audio task.
+ */
+void Sched_SendAudioCancelMsg(SchedContext* sched) {
+    osSendMesg(&sched->interruptQ, RDP_AUDIO_CANCEL_MSG, OS_MESG_BLOCK);
+}
+
+/**
+ * Sends a message to the scheduler to inform it that it should attempt
+ * to stop the last dispatched gfx task.
+ */
+void Sched_SendGfxCancelMsg(SchedContext* sched) {
+    osSendMesg(&sched->interruptQ, RSP_GFX_CANCEL_MSG, OS_MESG_BLOCK);
+}
+
+/**
+ * Fault Client for the scheduler. Reports information about the state of the scheduler
+ * and any current tasks in the crash debugger.
+ */
+void Sched_FaultClient(void* param1, void* param2) {
```

Comment:
```suggestion
void Sched_FaultClient(void* context, void* param2) {
```
or something like that? Why not type the argument?

Reply:
This is a fault client callback, the args are deliberately left generic much like the final arg of an override or post limb draw function in the skelanime system, since this callback can be passed any two arguments. Due to this, warnings are raised if I change the type to be more specific than `void*`, hence the cast. The casts in other callback functions with generic arguments have been known to affect codegen also, so I feel it's best to keep it consistent even when it doesn't matter.